VIPs/vip-10.md

```diff
@@ -0,0 +1,109 @@
+---
+vip: 10
+title: Storage node registration (on-chain)
+author: Chris Were <chris@verida.io>
+discussions-to: #dev-protocol on Verida Discord
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-03-22
+---
+
+# Overview
+
+The selection of storage nodes occurs when a DID is first created and when a context is connected to for the first time. The storage nodes are used to store the actual DID Documents and the same nodes are typically (but not necessarily) used to store data for each application context.
+
+This VIP provides a specification for storing the available storage nodes on-chain.
+
+This is phase 2 of [VIP-7: Storage node discovery and selection (off-chain)](./vip-7.md)
+
+This VIP does not implement any token based staking or slashing of nodes, it's purely focused on node addition, removal and discovery.
+
+# Motivation
+
+The current storage node database is centralized and needs to be stored on-chain to decentralize the discovery and selection of nodes on the network.
+
+# Specification
+
+## Data structures
+
+### StorageNode
+
+- didAddress: string
+- endpointUri: string
+- countryCode: string
+- regionCode: string
+- datacenterId: id
+
+### Datacenter
+
+- id: int
+- countryCode: string
+- regionCode: string
+- lat: int
+- long: int
+
+## State variables
+
+TBA
```

Comment:
@ITStar10 Can you propose appropriate state variables

Reply:
- `countryCode` and `regionCode` are duplicated between StorageNode & Datacenter. Or are tehy different values?

Reply:
This is the draft of state variables:
```
struct StorageNode {
  address didAddress;
  string endpointUri;
  uint datacenterId;

  uint unregisterTime;
}

struct Datacenter {
  string countryCode;
  string regionCode;
  uint lat;
  uint long;
}

/** mapping of storageNode ID => Storage*/
map(uint => StorageNode) private _storageNodeMap;
/** mapping of did address => storageNode ID*/
map(address => uint) private _addressNodeMap;
/** mapping of endpoint uri => storageNode ID*/
map(string => uint) private _endpointNodeMap;
/** mapping of countryCode => StorgeNode ID list*/
map(string => EnumerableSet.UintSet) private _countryNodes;
/** mapping of regionCode => StorageNode ID list*/
map(string => EnumerableSet.UintSet) private _regionNodes;

/** mapping of datacenterId => DataCenter */
map(uint => Datacenter) dataCenterMap
/** mapping of datacenterID => StorageNode ID list*/
map(uint => EnumerableSet.UintSet) private _dataCenterToNodeIdsmap;


CountersUpgradeable.Counter private _nodeIdCounter;
CountersUpgradeable.Counter private _dataCenterIdCounter;
```

Reply:
>countryCode and regionCode are duplicated between StorageNode & Datacenter. Or are tehy different values?

Correct, duplicated.

If you have them in your mapping it's okay to only store them in `Datacenter`.

Reply:
@ITStar10 Can you push a commit to add your proposed state variables?

Reply:
It might be better to push the State Variables after contract development done.

Reply:
@ITStar10 This is a draft, so put what you have in here for now and we will update at the end.